atomix/cluster/src/test/java/io/atomix/raft/RaftRule.java

```diff
@@ -280,43 +282,50 @@ public void doSnapshot(final long index) throws Exception {
   public boolean allNodesHaveSnapshotWithIndex(final long index) {
     return servers.values().stream()
             .map(RaftServer::getContext)
-            .map(RaftContext::getSnapshotStore)
-            .map(SnapshotStore::getCurrentSnapshotIndex)
+            .map(RaftContext::getPersistedSnapshotStore)
+            .map(PersistedSnapshotStore::getCurrentSnapshotIndex)
             .filter(idx -> idx == index)
             .count()
         == servers.values().size();
   }
 
-  public Snapshot getSnapshotFromLeader() {
+  public PersistedSnapshot getSnapshotFromLeader() {
     final var leader = getLeader().orElseThrow();
     final var context = leader.getContext();
-    final var snapshotStore = context.getSnapshotStore();
-    return snapshotStore.getCurrentSnapshot();
+    final var snapshotStore = context.getPersistedSnapshotStore();
+    return snapshotStore.getLatestSnapshot().orElseThrow();
   }
 
-  public Snapshot getSnapshotOnNode(final String nodeId) {
+  public PersistedSnapshot getSnapshotOnNode(final String nodeId) {
     final var raftServer = servers.get(nodeId);
     final var context = raftServer.getContext();
-    final var snapshotStore = context.getSnapshotStore();
-    return snapshotStore.getCurrentSnapshot();
-  }
-
-  private void writeSnapshot(final long index, final long term, final SnapshotStore snapshotStore)
-      throws IOException {
-    final var dirName = index + "-snapshot";
-    final var snapshotDir = new File(snapshotStore.getPath().toFile(), dirName);
-    if (!snapshotDir.mkdirs()) {
-      throw new IllegalStateException("Was not able to create directory: " + snapshotDir.getName());
-    }
-
-    final var snapshotFile = new File(snapshotDir, "snapshot.file");
-    Files.write(
-        snapshotFile.toPath(),
-        RandomStringUtils.random(128).getBytes(),
-        StandardOpenOption.CREATE_NEW,
-        StandardOpenOption.WRITE);
-
-    snapshotStore.newSnapshot(index, term, new WallClockTimestamp(), snapshotDir.toPath());
+    final var snapshotStore = context.getPersistedSnapshotStore();
+    return snapshotStore.getLatestSnapshot().orElseThrow();
+  }
+
+  private void writeSnapshot(
+      final long index, final long term, final PersistedSnapshotStore persistedSnapshotStore) {
+
+    final var transientSnapshot =
+        persistedSnapshotStore.newTransientSnapshot(index, term, new WallClockTimestamp());
+    transientSnapshot.take(
+        path -> {
+          IoUtil.ensureDirectoryExists(path.toFile(), "snapshot dir should exist");
+          final var snapshotFile = path.resolve("snapshot.file");
+          try {
+            Files.write(
+                snapshotFile,
+                RandomStringUtils.random(128).getBytes(),
+                StandardOpenOption.CREATE_NEW,
+                StandardOpenOption.WRITE);
+          } catch (final IOException e) {
+            e.printStackTrace();
```

Comment:
That doesn't feel right. Don't we have a logger somewhere?

Reply:
Be aware that this is in a test